Empty tree_builder output fails with EXPECTED_BUILT_REPO_TREE

### Describe the bug

A `tree_builder` script that returns an empty file list fails evaluation instead of being treated as a valid empty tree.

### Steps to reproduce

1. `.terrateam/config.yml`:

```yaml
tree_builder:
  enabled: true
  script: |
    echo '[]'
```

2. Open a pull request and run `terrateam plan`.
3. The build-tree phase succeeds; the build-config phase never starts.

Server log:

```
[INFO]  ... : REPO_TREE : STORE : branch_ref=<sha>
[ERROR] ... : EXPECTED_BUILT_REPO_TREE
[INFO]  ... : PUBLISH : UNEXPECTED_TEMPORARY_ERR
```

The PR receives repeated "There was an unexpected error processing this operation" comments.

### Expected behavior

An empty tree is a valid result. Evaluation should continue with zero files.

### Actual behavior

Evaluation fails with `EXPECTED_BUILT_REPO_TREE` (or `EXPECTED_BUILT_REPO_TREE_DEST_BRANCH` when the config is on the destination branch).

Root cause: an empty tree stores zero rows in `repo_trees`, and `query_repo_tree` maps zero rows to `None`, which is indistinguishable from a tree that was never built. `built_repo_tree_branch` / `built_repo_tree_dest_branch` in `terrat_vcs_event_evaluator2_tasks.ml` treat `None` as a hard error.

A secondary effect: because the empty tree is never found in the database, `wm_sm_repo_tree.create` re-runs build-tree on every event rather than reusing the cached result.

### Terrateam version & environment details

Reproduced on the `terrat-ee-dev` images built from `main` as of 2026-07-21. GitHub VCS. Affects the v2 event evaluator; the v1 evaluator degraded gracefully via `CCOption.map_or`.

### Additional context

Workaround: emit one inert path (e.g. `[{"path": "README.md", "id": "x"}]`) instead of `[]`. A path matching no dirspace is harmless and makes the query return a result.

Note that a *failed* tree-build script reaches the same code path, so any fix must not treat "no rows" as "empty tree" unconditionally — that would silently swallow genuine build failures.

--- a/code/src/terrat/migrations/2026-07-21-add-repo-tree-builds.sql
+++ b/code/src/terrat/migrations/2026-07-21-add-repo-tree-builds.sql
@@ -1,0 +1,18 @@
+-- Records that a repo tree was successfully built for a commit.
+--
+-- The rows in [repo_trees] cannot distinguish "the tree was built and it is
+-- empty" from "the tree was never built", because an empty tree stores zero
+-- rows.  A tree builder that legitimately outputs no files was therefore
+-- treated as a missing tree and failed the evaluation.  This table records the
+-- build itself so the two cases can be told apart.
+create table repo_tree_builds (
+    sha text not null,
+    installation uuid not null,
+    created_at timestamp with time zone not null default (now()),
+    primary key (installation, sha)
+);
+
+-- Backfill the trees we already have so existing commits are not rebuilt.
+insert into repo_tree_builds (installation, sha)
+select distinct installation, sha from repo_trees
+on conflict do nothing;

--- a/code/src/terrat/terrat_migrations.ml
+++ b/code/src/terrat/terrat_migrations.ml
@@ -280,6 +280,7 @@ let migrations =
     ( "extend-repo-configs-with-history",
       run_sql [%blob "migrations/2026-07-05-extend-repo-configs-with-history.sql"] );
     ("add-tasks-user-id", run_sql [%blob "migrations/2026-07-23-add-tasks-user-id.sql"]);
+    ("add-repo-tree-builds", run_sql [%blob "migrations/2026-07-21-add-repo-tree-builds.sql"]);
   ]
 
 let run config storage = Mig.run { Migrate.config; storage; tx = () } migrations

--- a/code/src/terrat_vcs_service_github/sql/insert_repo_tree_build.sql
+++ b/code/src/terrat_vcs_service_github/sql/insert_repo_tree_build.sql
@@ -1,0 +1,8 @@
+insert into repo_tree_builds (sha, installation)
+select
+        $sha,
+        gim.core_id
+from github_installations_map as gim
+where gim.installation_id = $installation_id
+on conflict on constraint repo_tree_builds_pkey
+do nothing

--- a/code/src/terrat_vcs_service_github/sql/select_repo_tree_build.sql
+++ b/code/src/terrat_vcs_service_github/sql/select_repo_tree_build.sql
@@ -1,0 +1,5 @@
+select rtb.sha
+from repo_tree_builds as rtb
+inner join github_installations_map as gim
+      on gim.core_id = rtb.installation
+where gim.installation_id = $installation_id and rtb.sha = $sha

--- a/code/src/terrat_vcs_service_github/terrat_vcs_service_github_provider.ml
+++ b/code/src/terrat_vcs_service_github/terrat_vcs_service_github_provider.ml
@@ -357,6 +357,13 @@ module Db = struct
         /% Var.(array (option (boolean "changed")))
         /% Var.(str_array (option (text "id"))))
 
+    let insert_repo_tree_build =
+      Pgsql_io.Typed_sql.(
+        sql
+        /^ read [%blob "sql/insert_repo_tree_build.sql"]
+        /% Var.bigint "installation_id"
+        /% Var.text "sha")
+
     let upsert_flow_state_query = read [%blob "sql/update_flow_state.sql"]
 
     let upsert_flow_state () =
@@ -454,6 +461,16 @@ module Db = struct
         /% Var.bigint "installation_id"
         /% Var.text "sha"
         /% Var.(option (text "base_sha")))
+
+    let select_repo_tree_build =
+      Pgsql_io.Typed_sql.(
+        sql
+        //
+        (* sha *)
+        Ret.text
+        /^ read [%blob "sql/select_repo_tree_build.sql"]
+        /% Var.bigint "installation_id"
+        /% Var.text "sha")
 
     let select_next_work_manifest =
       Pgsql_io.Typed_sql.(
@@ -1267,7 +1284,22 @@ module Db = struct
               (CCList.map (fun { I.id; _ } -> id) chunk)))
       (CCList.chunks not_a_bad_chunk_size files)
     >>= function
-    | Ok () -> Abb.Future.return (Ok ())
+    | Ok () -> (
+        (* Record the build itself.  An empty tree stores no rows, so this is
+           the only way to tell a built-but-empty tree apart from one that was
+           never built. *)
+        Metrics.Psql_query_time.time (Metrics.psql_query_time "insert_repo_tree_build") (fun () ->
+            Pgsql_io.Prepared_stmt.execute
+              db
+              Sql.insert_repo_tree_build
+              (CCInt64.of_int @@ Api.Account.id account)
+              (Api.Ref.to_string ref_))
+        >>= function
+        | Ok () -> Abb.Future.return (Ok ())
+        | Error (#Pgsql_io.err as err) ->
+            Prmths.Counter.inc_one Metrics.pgsql_errors_total;
+            Logs.err (fun m -> m "%s : ERROR : %a" request_id Pgsql_io.pp_err err);
+            Abb.Future.return (Error `Error))
     | Error (#Pgsql_io.err as err) ->
         Prmths.Counter.inc_one Metrics.pgsql_errors_total;
         Logs.err (fun m -> m "%s : ERROR : %a" request_id Pgsql_io.pp_err err);
@@ -1651,7 +1683,23 @@ module Db = struct
           (Api.Ref.to_string ref_)
           (CCOption.map Api.Ref.to_string base_ref))
     >>= function
-    | Ok [] -> Abb.Future.return (Ok None)
+    | Ok [] -> (
+        (* No rows is ambiguous: either the tree was built and is empty, or it
+           was never built.  [repo_tree_builds] tells the two apart. *)
+        Metrics.Psql_query_time.time (Metrics.psql_query_time "select_repo_tree_build") (fun () ->
+            Pgsql_io.Prepared_stmt.fetch
+              db
+              Sql.select_repo_tree_build
+              ~f:CCFun.id
+              (CCInt64.of_int @@ Api.Account.id account)
+              (Api.Ref.to_string ref_))
+        >>= function
+        | Ok (_ :: _) -> Abb.Future.return (Ok (Some []))
+        | Ok [] -> Abb.Future.return (Ok None)
+        | Error (#Pgsql_io.err as err) ->
+            Prmths.Counter.inc_one Metrics.pgsql_errors_total;
+            Logs.err (fun m -> m "%s : ERROR : %a" request_id Pgsql_io.pp_err err);
+            Abb.Future.return (Error `Error))
     | Ok files -> Abb.Future.return (Ok (Some files))
     | Error (#Pgsql_io.err as err) ->
         Prmths.Counter.inc_one Metrics.pgsql_errors_total;

--- a/code/src/terrat_vcs_service_gitlab/sql/insert_repo_tree_build.sql
+++ b/code/src/terrat_vcs_service_gitlab/sql/insert_repo_tree_build.sql
@@ -1,0 +1,8 @@
+insert into repo_tree_builds (sha, installation)
+select
+        $sha,
+        gim.core_id
+from gitlab_installations_map as gim
+where gim.installation_id = $installation_id
+on conflict on constraint repo_tree_builds_pkey
+do nothing

--- a/code/src/terrat_vcs_service_gitlab/sql/select_repo_tree_build.sql
+++ b/code/src/terrat_vcs_service_gitlab/sql/select_repo_tree_build.sql
@@ -1,0 +1,5 @@
+select rtb.sha
+from repo_tree_builds as rtb
+inner join gitlab_installations_map as gim
+      on gim.core_id = rtb.installation
+where gim.installation_id = $installation_id and rtb.sha = $sha

--- a/code/src/terrat_vcs_service_gitlab/terrat_vcs_service_gitlab_provider.ml
+++ b/code/src/terrat_vcs_service_gitlab/terrat_vcs_service_gitlab_provider.ml
@@ -357,6 +357,13 @@ module Db = struct
         /% Var.(array (option (boolean "changed")))
         /% Var.(str_array (option (text "id"))))
 
+    let insert_repo_tree_build =
+      Pgsql_io.Typed_sql.(
+        sql
+        /^ read [%blob "sql/insert_repo_tree_build.sql"]
+        /% Var.bigint "installation_id"
+        /% Var.text "sha")
+
     let upsert_flow_state_query = read [%blob "sql/update_flow_state.sql"]
 
     let upsert_flow_state () =
@@ -454,6 +461,16 @@ module Db = struct
         /% Var.bigint "installation_id"
         /% Var.text "sha"
         /% Var.(option (text "base_sha")))
+
+    let select_repo_tree_build =
+      Pgsql_io.Typed_sql.(
+        sql
+        //
+        (* sha *)
+        Ret.text
+        /^ read [%blob "sql/select_repo_tree_build.sql"]
+        /% Var.bigint "installation_id"
+        /% Var.text "sha")
 
     let select_next_work_manifest =
       Pgsql_io.Typed_sql.(
@@ -1266,7 +1283,22 @@ module Db = struct
               (CCList.map (fun { I.id; _ } -> id) chunk)))
       (CCList.chunks not_a_bad_chunk_size files)
     >>= function
-    | Ok () -> Abb.Future.return (Ok ())
+    | Ok () -> (
+        (* Record the build itself.  An empty tree stores no rows, so this is
+           the only way to tell a built-but-empty tree apart from one that was
+           never built. *)
+        Metrics.Psql_query_time.time (Metrics.psql_query_time "insert_repo_tree_build") (fun () ->
+            Pgsql_io.Prepared_stmt.execute
+              db
+              Sql.insert_repo_tree_build
+              (CCInt64.of_int @@ Api.Account.id account)
+              (Api.Ref.to_string ref_))
+        >>= function
+        | Ok () -> Abb.Future.return (Ok ())
+        | Error (#Pgsql_io.err as err) ->
+            Prmths.Counter.inc_one Metrics.pgsql_errors_total;
+            Logs.err (fun m -> m "%s : ERROR : %a" request_id Pgsql_io.pp_err err);
+            Abb.Future.return (Error `Error))
     | Error (#Pgsql_io.err as err) ->
         Prmths.Counter.inc_one Metrics.pgsql_errors_total;
         Logs.err (fun m -> m "%s : ERROR : %a" request_id Pgsql_io.pp_err err);
@@ -1607,7 +1639,23 @@ module Db = struct
           (Api.Ref.to_string ref_)
           (CCOption.map Api.Ref.to_string base_ref))
     >>= function
-    | Ok [] -> Abb.Future.return (Ok None)
+    | Ok [] -> (
+        (* No rows is ambiguous: either the tree was built and is empty, or it
+           was never built.  [repo_tree_builds] tells the two apart. *)
+        Metrics.Psql_query_time.time (Metrics.psql_query_time "select_repo_tree_build") (fun () ->
+            Pgsql_io.Prepared_stmt.fetch
+              db
+              Sql.select_repo_tree_build
+              ~f:CCFun.id
+              (CCInt64.of_int @@ Api.Account.id account)
+              (Api.Ref.to_string ref_))
+        >>= function
+        | Ok (_ :: _) -> Abb.Future.return (Ok (Some []))
+        | Ok [] -> Abb.Future.return (Ok None)
+        | Error (#Pgsql_io.err as err) ->
+            Prmths.Counter.inc_one Metrics.pgsql_errors_total;
+            Logs.err (fun m -> m "%s : ERROR : %a" request_id Pgsql_io.pp_err err);
+            Abb.Future.return (Error `Error))
     | Ok files -> Abb.Future.return (Ok (Some files))
     | Error (#Pgsql_io.err as err) ->
         Prmths.Counter.inc_one Metrics.pgsql_errors_total;

--- a/docs/src/content/docs/reference/configuration/tree-builder.mdx
+++ b/docs/src/content/docs/reference/configuration/tree-builder.mdx
@@ -29,30 +29,33 @@ When `tree_builder` is enabled, Terrateam executes your custom script during the
 
 ### Output
 
-The script must output valid JSON to stdout with the following structure
+The script must output a JSON array of file objects to stdout
 
 ```json
-{
-  "files": [
-    {
-      "path": "path/to/file1.tf",
-      "id": "abc123def456"
-    },
-    {
-      "path": "path/to/file2.tf",
-      "id": "789ghi012jkl"
-    }
-  ]
-}
+[
+  {
+    "path": "path/to/file1.tf",
+    "id": "abc123def456"
+  },
+  {
+    "path": "path/to/file2.tf",
+    "id": "789ghi012jkl"
+  }
+]
 ```
+
+The output is a bare array, not an object wrapping a `files` key. A script that
+outputs `{"files": [...]}` fails with `TypeError: string indices must be
+integers`.
+
+An empty array is valid and means the repository has no files to track.
 
 ### Output Fields
 
 | Field | Type | Required | Description |
 | --- | --- | --- | --- |
-| files | Array | Yes | Array of file objects |
-| files[].path | String | Yes | Relative path from repository root |
-| files[].id | String | Yes | Unique identifier for the file (typically a hash). Terrateam compares IDs between branches to detect changes |
+| [].path | String | Yes | Relative path from repository root |
+| [].id | String | No | Unique identifier for the file (typically a hash). Terrateam compares IDs between branches to detect changes |
 
 ## How It Works
 

